### PR TITLE
Optimize constructors of matrix and vector

### PR DESCRIPTION
**Pull request by toor1245**

### CODE REVIEW

**Comment by Dilorfin** [comment]
I think it would be great if Functional Testing project could be ran as usual tests

**Comment by Dilorfin** [comment]
Are you sure that this code written with Unsafe library is better then creating specified generic for each supported type?

**Comment by toor1245** [comment]
> Are you sure that this code written with Unsafe library is better then creating specified generic for each supported type?

I except to use Unsafe library instead of generic type in cases where operation will be same data type and use `MathGeneric` for different data types. Also, in my point of vision I think will need eliminate `MathExtension` class because we have significant penalty of performance compared `MathGeneric`, thus the better way will replace all methods of `MathExtension` to `MathGeneric` and marked  `MathExtension` class like Obsolete

**Comment by toor1245** [comment]
> I think it would be great if Functional Testing project could be ran as usual tests

Actually I don`t know other way how to run functional tests without entry point for benchmarking and I created namespace threading for checks how will lead static field `Dictionary` and `delegate` in multithreading, by reason that in unit testing not show error when occurs 'race data', so this way need to run manually. By the way, tests showed that no 'race data' with the 'delegate' was detected

**Review by Dilorfin — CHANGES_REQUESTED** [review]

File: tests/MatrixDotNet.FunctionalTesting/BenchMatrix/BenchCtorUnsafeVsDefault/logs/MatrixDotNet.FunctionalTesting.BenchMatrix.BenchCtorUnsafeAndDefault-20201031-045010.log [hunk 1]
```diff
@@ -0,0 +1,429 @@
+// Validating benchmarks:
```
[the remark this answers is not in the document]
Comment (toor1245) [reply tests/MatrixDotNet.FunctionalTesting/BenchMatrix/BenchCtorUnsafeVsDefault/logs/MatrixDotNet.FunctionalTesting.BenchMatrix.BenchCtorUnsafeAndDefault-20201031-045010.log:1]: > May be we should add log to .gitignore?
Yes, I fully agree with your opinion.

[the remark this answers is not in the document]
Comment (toor1245) [reply tests/MatrixDotNet.FunctionalTesting/BenchMatrix/BenchCtorUnsafeVsDefault/logs/MatrixDotNet.FunctionalTesting.BenchMatrix.BenchCtorUnsafeAndDefault-20201031-045010.log:1]: > > May be we should add log to .gitignore?

Yes, I fully agree with your opinion.

[the remark this answers is not in the document]
Comment (toor1245) [reply tests/MatrixDotNet.FunctionalTesting/BenchMatrix/BenchCtorUnsafeVsDefault/logs/MatrixDotNet.FunctionalTesting.BenchMatrix.BenchCtorUnsafeAndDefault-20201031-045010.log]: I added in .gitignore file which will ignore logs directories

**Review by Dilorfin — APPROVED** [review]

**Pull request merged**